File: Little Artist/Views/Children/ShareManagementView.swift
//
//  ShareManagementView.swift
//  Little Artist
//
//  Sharing management view — displays participants fetched from Firestore,
//  allows the owner to send invite links, stop sharing, and participants to leave.
//

import SwiftUI
import SwiftData

/// A sharing management view that shows participants from Firestore.
///
/// A branded sharing experience powered by ``FirestoreRepository``.
struct ShareManagementView: View {
    let child: Child
    let shareId: String
    /// When true, the view was just created — show the share link immediately.
    var isNewShare: Bool = false

    var onStoppedSharing: (() -> Void)? = nil

    @Environment(\.dismiss) private var dismiss
    @Environment(\.modelContext) private var modelContext

    @State private var participants: [(userId: String, role: String, acceptedAt: Date?)] = []
    @State private var isLoadingParticipants = true
    @State private var sharePayload: SharePayload?
    @State private var isStoppingShare = false
    @State private var isLeavingShare = false
    @State private var stopError: String?

    private var isCurrentUserOwner: Bool {
        !child.isShared
    }

    /// The share code that the recipient enters in their app to join.
    private var shareCode: String { shareId }

    // MARK: - Body

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: Brand.sectionSpacing) {
                    // Child avatar header
                    childHeader
                        .padding(.top, 16)

                    // People section
                    peopleSection

                    // Share options
                    shareOptionsSection

                    // Owner: stop sharing, Participant: leave
                    if isCurrentUserOwner {
                        stopSharingButton
                    } else {
                        leaveProfileButton
                    }
                }
                .padding(.horizontal, Brand.screenPadding)
                .padding(.bottom, 32)
            }
            .background(BrandAppBackground())
            .navigationTitle("Sharing")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { dismiss() }
                        .fontWeight(.semibold)
                }
            }
            .task {
                await loadParticipants()
                // If this is a brand-new share, show the share sheet immediately
                if isNewShare {
                    presentShareLink()
                }
            }
            .sheet(item: $sharePayload) { payload in
                ActivityView(activityItems: payload.items)
            }
            .alert("Error", isPresented: Binding(
                get: { stopError != nil },
                set: { if !$0 { stopError = nil } }
            )) {
                Button("OK") { stopError = nil }
            } message: {
                Text(stopError ?? "")
            }
        }
    }

    // MARK: - Child Header

    private var childHeader: some View {
        VStack(spacing: 12) {
            // Circular child avatar
            Group {
                if let data = child.avatarImageData,
                   let uiImage = UIImage(data: data) {
                    Image(uiImage: uiImage)
                        .resizable()
                        .scaledToFill()
                        .frame(width: 80, height: 80)
                        .clipShape(Circle())
                } else {
                    Circle()
                        .fill(Color(hex: child.avatarColor))
                        .frame(width: 80, height: 80)
                        .overlay {
                            Text(String(child.name.prefix(1)).uppercased())
                                .font(.system(size: 34, weight: .bold, design: .rounded))
                                .foregroundStyle(.white)
                        }
                }
            }
            .brandCardShadow()

            Text(child.name)
                .font(Brand.title2Font)
                .foregroundStyle(Brand.charcoal)
        }
    }

    // MARK: - People Section

    private var peopleSection: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text("People")
                .font(Brand.captionFont)
                .foregroundStyle(Brand.warmGray)
                .textCase(.uppercase)
                .padding(.leading, 4)
                .padding(.bottom, 8)

            VStack(spacing: 0) {
                // Owner row
                ownerRow
                    .padding(.horizontal, 16)
                    .padding(.vertical, 14)

                if isLoadingParticipants {
                    Divider().padding(.leading, 56)
                    HStack(spacing: 12) {
                        ProgressView()
                        Text("Loading participants...")
                            .font(Brand.captionFont)
                            .foregroundStyle(Brand.warmGray)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 14)
                }

                // Other participants
                ForEach(participants, id: \.userId) { participant in
                    Divider().padding(.leading, 56)
                    participantRow(participant)
                        .padding(.horizontal, 16)
                        .padding(.vertical, 14)
                }

                // Only owners can invite more people
                if isCurrentUserOwner {
                    Divider().padding(.leading, 56)

                    Button {
                        presentShareLink()
                    } label: {
                        HStack(spacing: 12) {
                            Image(systemName: "plus.circle.fill")
                                .font(.system(size: 28))
                                .foregroundStyle(Brand.primary)
                            Text("Share With More People")
                                .font(Brand.bodyFont)
                                .foregroundStyle(Brand.primary)
                        }
                        .padding(.horizontal, 16)
                        .padding(.vertical, 14)
                    }
                }
            }
            .background(Brand.surface)
            .clipShape(RoundedRectangle(cornerRadius: Brand.radiusCard))
            .brandCardShadow()
        }
    }

    private var ownerRow: some View {
        HStack(spacing: 12) {
            Image(systemName: "person.crop.circle.fill")
                .font(.system(size: 28))
                .foregroundStyle(Brand.primary)

            VStack(alignment: .leading, spacing: 2) {
                Text(isCurrentUserOwner ? "You" : "Owner")
                    .font(Brand.bodyFont)
                    .foregroundStyle(Brand.charcoal)
                Text("Owner")
                    .font(Brand.caption2Font)
                    .foregroundStyle(Brand.warmGray)
            }

            Spacer()
        }
    }

    private func participantRow(_ participant: (userId: String, role: String, acceptedAt: Date?)) -> some View {
        HStack(spacing: 12) {
            Image(systemName: "person.crop.circle")
                .font(.system(size: 28))
                .foregroundStyle(Brand.sky)

            VStack(alignment: .leading, spacing: 2) {
                Text(participant.userId.prefix(8) + "...")
                    .font(Brand.bodyFont)
                    .foregroundStyle(Brand.charcoal)
                Text(participantRoleLabel(participant.role))
                    .font(Brand.caption2Font)
                    .foregroundStyle(Brand.warmGray)
            }

            Spacer()

            if participant.acceptedAt == nil {
                Text("Pending")
                    .font(Brand.caption2Font)
                    .foregroundStyle(Brand.primary)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(Brand.primaryTint)
                    .clipShape(Capsule())
            }
        }
    }

    // MARK: - Share Options

    private var shareOptionsSection: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text("Share Options")
                .font(Brand.captionFont)
                .foregroundStyle(Brand.warmGray)
                .textCase(.uppercase)
                .padding(.leading, 4)
                .padding(.bottom, 8)

            VStack(spacing: 0) {
                HStack {
                    Image(systemName: "pencil.circle.fill")
                        .font(.system(size: 22))
                        .foregroundStyle(Brand.sage)
                    Text("People you invite can make changes and add others.")
                        .font(Brand.captionFont)
                        .foregroundStyle(Brand.charcoal)
                }
                .padding(16)
            }
            .background(Brand.surface)
            .clipShape(RoundedRectangle(cornerRadius: Brand.radiusCard))
            .brandCardShadow()
        }
    }

    // MARK: - Stop Sharing

    private var stopSharingButton: some View {
        Button(role: .destructive) {
            Task { await stopSharing() }
        } label: {
            HStack {
                if isStoppingShare {
                    ProgressView()
                        .tint(Brand.dustyRose)
                } else {
                    Text("Stop Sharing")
                        .font(Brand.headlineFont)
                }
            }
            .foregroundStyle(Brand.dustyRose)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 16)
            .background(Brand.surface)
            .clipShape(RoundedRectangle(cornerRadius: Brand.radiusCard))
            .brandCardShadow()
        }
        .disabled(isStoppingShare)
    }

    // MARK: - Leave Profile (Participant)

    private var leaveProfileButton: some View {
        Button(role: .destructive) {
            Task { await leaveProfile() }
        } label: {
            HStack {
                if isLeavingShare {
                    ProgressView()
                        .tint(Brand.dustyRose)
                } else {
                    Image(systemName: "person.badge.minus")
                        .font(.system(size: 16, weight: .medium))
                    Text("Leave Profile")
                        .font(Brand.headlineFont)
                }
            }
            .foregroundStyle(Brand.dustyRose)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 16)
            .background(Brand.surface)
            .clipShape(RoundedRectangle(cornerRadius: Brand.radiusCard))
            .brandCardShadow()
        }
        .disabled(isLeavingShare)
    }

    // MARK: - Helpers

    private func participantRoleLabel(_ role: String) -> String {
        switch role {
        case "editor": return "Can make changes"
        case "viewer": return "View only"
        default: return "Participant"
        }
    }

    // MARK: - Actions

    private func loadParticipants() async {
        isLoadingParticipants = true
        do {
            participants = try await FirestoreRepository.shared.fetchParticipants(shareId: shareId)
        } catch {
            // Silently fail — empty participants list is fine
        }
        isLoadingParticipants = false
    }

    private func presentShareLink() {
        let text = "Join me on Artling to see \(child.name)'s artwork! Open the app, go to Settings → Join Shared Profile, and enter this code:\n\n\(shareCode)"
        sharePayload = SharePayload(items: [text])
    }

    private func stopSharing() async {
        isStoppingShare = true
        do {
            try await FirestoreRepository.shared.stopSharing(shareId: shareId)
            dismiss()
            onStoppedSharing?()
        } catch {
            stopError = error.localizedDescription
            isStoppingShare = false
        }
    }

    private func leaveProfile() async {
        isLeavingShare = true
        do {
            try await FirestoreRepository.shared.leaveShare(shareId: shareId)
            // Remove local mirror of shared child
            modelContext.delete(child)
            dismiss()
            onStoppedSharing?()
        } catch {
            stopError = error.localizedDescription
            isLeavingShare = false
        }
    }
}

File: little_artist_flutter/lib/views/children/share_management_view.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:share_plus/share_plus.dart';

import '../../models/database.dart';
import '../../providers/database_provider.dart';
import '../../providers/sync_provider.dart';
import '../../services/firestore_repository.dart';
import '../../utils/brand_tokens.dart';

/// Sharing management flow copied from the SwiftUI structure.
class ShareManagementView extends ConsumerStatefulWidget {
  const ShareManagementView({
    super.key,
    required this.child,
    required this.shareId,
    this.isNewShare = false,
    this.onStoppedSharing,
    this.onLeftProfile,
  });

  final Child child;
  final String shareId;
  final bool isNewShare;
  final Future<void> Function()? onStoppedSharing;
  final VoidCallback? onLeftProfile;

  @override
  ConsumerState<ShareManagementView> createState() =>
      _ShareManagementViewState();
}

class _ShareManagementViewState extends ConsumerState<ShareManagementView> {
  List<ShareParticipant> _participants = const [];
  bool _isLoadingParticipants = true;
  bool _isStoppingShare = false;
  bool _isLeavingProfile = false;

  bool get _isCurrentUserOwner => !widget.child.isShared;

  @override
  void initState() {
    super.initState();
    _loadParticipants();
  }

  Future<void> _loadParticipants() async {
    try {
      final participants = await ref
          .read(firestoreRepositoryProvider)
          .fetchParticipants(widget.shareId);
      if (!mounted) {
        return;
      }
      setState(() {
        _participants = participants;
        _isLoadingParticipants = false;
      });
      if (widget.isNewShare) {
        WidgetsBinding.instance.addPostFrameCallback(
          (_) => _presentShareLink(),
        );
      }
    } catch (_) {
      if (!mounted) {
        return;
      }
      setState(() => _isLoadingParticipants = false);
    }
  }

  Future<void> _presentShareLink() async {
    final text =
        "Join me on Artling to see ${widget.child.name}'s artwork! Open the app, go to Settings -> Join Shared Profile, and enter this code:\n\n${widget.shareId}";
    await Share.share(text);
  }

  Future<void> _stopSharing() async {
    if (_isStoppingShare) {
      return;
    }

    setState(() => _isStoppingShare = true);
    try {
      await ref.read(firestoreRepositoryProvider).stopSharing(widget.shareId);
      await ref
          .read(databaseProvider)
          .childDao
          .updateChild(widget.child.copyWith(isShared: false));
      await widget.onStoppedSharing?.call();
      if (mounted) {
        Navigator.of(context).pop();
      }
    } catch (e) {
      if (!mounted) {
        return;
      }
      setState(() => _isStoppingShare = false);
      ScaffoldMessenger.of(
        context,
      ).showSnackBar(SnackBar(content: Text('Error: $e')));
    }
  }

  Future<void> _leaveProfile() async {
    if (_isLeavingProfile) {
      return;
    }

    setState(() => _isLeavingProfile = true);
    try {
      await ref.read(firestoreRepositoryProvider).leaveShare(widget.shareId);
      await ref.read(databaseProvider).childDao.deleteChild(widget.child.id);
      widget.onLeftProfile?.call();
      if (mounted) {
        Navigator.of(context).pop();
      }
    } catch (e) {
      if (!mounted) {
        return;
      }
      setState(() => _isLeavingProfile = false);
      ScaffoldMessenger.of(
        context,
      ).showSnackBar(SnackBar(content: Text('Error: $e')));
    }
  }

  @override
  Widget build(BuildContext context) {
    final initial = widget.child.name.isEmpty
        ? '?'
        : widget.child.name[0].toUpperCase();
    final avatarColor = Color(
      int.parse('FF${widget.child.avatarColor}', radix: 16),
    );

    return Scaffold(
      backgroundColor: Brand.cream,
      appBar: AppBar(
        backgroundColor: Brand.cream,
        title: Text(
          'Sharing',
          style: Brand.title2Font.copyWith(color: Brand.charcoal),
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.of(context).pop(),
            child: Text(
              'Done',
              style: Brand.bodyFont.copyWith(
                color: Brand.charcoal,
                fontWeight: FontWeight.w600,
              ),
            ),
          ),
        ],
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.fromLTRB(20, 16, 20, 32),
        child: Column(
          children: [
            _buildChildHeader(initial, avatarColor),
            const SizedBox(height: 28),
            _buildPeopleSection(),
            const SizedBox(height: 24),
            _buildShareOptionsSection(),
            const SizedBox(height: 24),
            if (_isCurrentUserOwner)
              _buildDangerButton(
                label: 'Stop Sharing',
                isBusy: _isStoppingShare,
                onPressed: _stopSharing,
              )
            else
              _buildDangerButton(
                label: 'Leave Profile',
                isBusy: _isLeavingProfile,
                onPressed: _leaveProfile,
                icon: Icons.person_remove_alt_1_rounded,
              ),
          ],
        ),
      ),
    );
  }

  Widget _buildChildHeader(String initial, Color avatarColor) {
    return Column(
      children: [
        Container(
          width: 80,
          height: 80,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: avatarColor,
            boxShadow: Brand.cardShadow,
          ),
          child: widget.child.avatarImageData != null
              ? ClipOval(
                  child: Image.memory(
                    widget.child.avatarImageData!,
                    fit: BoxFit.cover,
                  ),
                )
              : Center(
                  child: Text(
                    initial,
                    style: Brand.displayFont.copyWith(
                      color: Colors.white,
                      fontSize: 34,
                    ),
                  ),
                ),
        ),
        const SizedBox(height: 12),
        Text(
          widget.child.name,
          style: Brand.title2Font.copyWith(color: Brand.charcoal),
        ),
      ],
    );
  }

  Widget _buildPeopleSection() {
    return _LabeledCardSection(
      title: 'People',
      child: Column(
        children: [
          _PersonRow(
            icon: Icons.account_circle_rounded,
            iconColor: Brand.primary,
            title: _isCurrentUserOwner ? 'You' : 'Owner',
            subtitle: 'Owner',
          ),
          if (_isLoadingParticipants)
            const _DividerWithInset()
          else
            const SizedBox.shrink(),
          if (_isLoadingParticipants)
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 14),
              child: Row(
                children: [
                  const SizedBox(
                    width: 18,
                    height: 18,
                    child: CircularProgressIndicator(strokeWidth: 2),
                  ),
                  const SizedBox(width: 12),
                  Text(
                    'Loading participants...',
                    style: Brand.captionFont.copyWith(color: Brand.warmGray),
                  ),
                ],
              ),
            ),
          for (final participant in _participants) ...[
            const _DividerWithInset(),
            _PersonRow(
              icon: Icons.account_circle_outlined,
              iconColor: Brand.sky,
              title:
                  '${participant.userId.substring(0, participant.userId.length.clamp(0, 8))}...',
              subtitle: _participantRoleLabel(participant.role),
              trailing: participant.acceptedAt == null
                  ? Container(
                      padding: const EdgeInsets.symmetric(
                        horizontal: 8,
                        vertical: 4,
                      ),
                      decoration: BoxDecoration(
                        color: Brand.primaryTint,
                        borderRadius: BorderRadius.circular(999),
                      ),
                      child: Text(
                        'Pending',
                        style: Brand.captionFont.copyWith(color: Brand.primary),
                      ),
                    )
                  : null,
            ),
          ],
          if (_isCurrentUserOwner) ...[
            const _DividerWithInset(),
            InkWell(
              onTap: _presentShareLink,
              child: Padding(
                padding: const EdgeInsets.symmetric(
                  horizontal: 16,
                  vertical: 14,
                ),
                child: Row(
                  children: [
                    const Icon(
                      Icons.add_circle_rounded,
                      color: Brand.primary,
                      size: 28,
                    ),
                    const SizedBox(width: 12),
                    Text(
                      'Share With More People',
                      style: Brand.bodyFont.copyWith(color: Brand.primary),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ],
      ),
    );
  }

  Widget _buildShareOptionsSection() {
    return _LabeledCardSection(
      title: 'Share Options',
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            const Icon(Icons.edit_note_rounded, color: Brand.sage, size: 22),
            const SizedBox(width: 12),
            Expanded(
              child: Text(
                'People you invite can make changes and add others.',
                style: Brand.captionFont.copyWith(color: Brand.charcoal),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildDangerButton({
    required String label,
    required VoidCallback onPressed,
    required bool isBusy,
    IconData? icon,
  }) {
    return SizedBox(
      width: double.infinity,
      child: ElevatedButton(
        onPressed: isBusy ? null : onPressed,
        style: ElevatedButton.styleFrom(
          backgroundColor: Brand.surface,
          foregroundColor: Brand.dustyRose,
          disabledBackgroundColor: Brand.surface,
          disabledForegroundColor: Brand.dustyRose,
          minimumSize: const Size.fromHeight(56),
          elevation: 0,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(Brand.radiusCard),
          ),
        ),
        child: isBusy
            ? const SizedBox(
                width: 20,
                height: 20,
                child: CircularProgressIndicator(
                  strokeWidth: 2,
                  color: Brand.dustyRose,
                ),
              )
            : Row(
                mainAxisAlignment: MainAxisAlignment.center,
                mainAxisSize: MainAxisSize.min,
                children: [
                  if (icon != null) ...[
                    Icon(icon, size: 18),
                    const SizedBox(width: 8),
                  ],
                  Text(
                    label,
                    style: Brand.headlineFont.copyWith(color: Brand.dustyRose),
                  ),
                ],
              ),
      ),
    );
  }

  String _participantRoleLabel(String role) {
    switch (role) {
      case 'editor':
        return 'Can make changes';
      case 'viewer':
        return 'View only';
      default:
        return 'Participant';
    }
  }
}

class _LabeledCardSection extends StatelessWidget {
  const _LabeledCardSection({required this.title, required this.child});

  final String title;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 4, bottom: 8),
          child: Text(
            title,
            style: Brand.captionFont.copyWith(
              color: Brand.warmGray,
              fontWeight: FontWeight.w700,
              letterSpacing: 0.8,
            ),
          ),
        ),
        Container(
          decoration: BoxDecoration(
            color: Brand.surface,
            borderRadius: BorderRadius.circular(Brand.radiusCard),
            boxShadow: Brand.cardShadow,
          ),
          clipBehavior: Clip.antiAlias,
          child: child,
        ),
      ],
    );
  }
}

class _PersonRow extends StatelessWidget {
  const _PersonRow({
    required this.icon,
    required this.iconColor,
    required this.title,
    required this.subtitle,
    this.trailing,
  });

  final IconData icon;
  final Color iconColor;
  final String title;
  final String subtitle;
  final Widget? trailing;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 14),
      child: Row(
        children: [
          Icon(icon, color: iconColor, size: 28),
          const SizedBox(width: 12),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  title,
                  style: Brand.bodyFont.copyWith(color: Brand.charcoal),
                ),
                const SizedBox(height: 2),
                Text(
                  subtitle,
                  style: Brand.captionFont.copyWith(color: Brand.warmGray),
                ),
              ],
            ),
          ),
          ...switch (trailing) {
            final Widget widget => [widget],
            null => const <Widget>[],
          },
        ],
      ),
    );
  }
}

class _DividerWithInset extends StatelessWidget {
  const _DividerWithInset();

  @override
  Widget build(BuildContext context) {
    return const Padding(
      padding: EdgeInsets.only(left: 56),
      child: Divider(height: 1, color: Brand.softTan),
    );
  }
}
